Automattic/wp-calypso
# Use babel to transpile TS projects

### Summary of changes

* Use `babel` to transpile TypeScript packages at bundle time.
* Use `tsc` to check types and generate type definitions.
* Use [build mode](https://www.typescriptlang.org/docs/handbook/project-references.html#build-mode-for-typescript) to generate types for all packages quickly.

Inspired by the discussion in #47002 by @jsnajdr and @sirreal 

### Background

Our Babel config can [transpile TypeScript](https://github.com/Automattic/wp-calypso/blob/master/packages/calypso-build/babel/default.js#L27) files. However we are not using that capability. Instead, we have a script `build-packages` that runs `tsc` on each package, creating an ESM compilation. Then that compilation is read by webpack and bundled with the rest of the app. This is not optimal because TypeScript can't do a smarter transpilation based on the target (evergreen vs fallback), and also it wastes time as packages are transpiled twice.

### Changes

* Exposes the untranspiled TypeScript code in TS packages using the property `calypso:src` that we introduced in #44824. This allows Webpack and babel to read the original TypeScript code and transpile it.

* Create a `build` step on each TypeScript package that will generate CJS, ESM and types. This is used on `prepack`, when the package is being built for publishing.

* Create a "main" `tsconfig.json` that references all TS packages, used as an entrypoint for `tsc`. 

### Notes

* We don't really need to run `tsc` at all for Calypso to compile and work. It is all handled by babel. The only reason to run `tsc` at installation time is to generate type definitions that can be used by the IDE.

* In theory we could use just `tsc --emitDeclarationOnly` to generate the definition files. However, `tsc --build` also generates the types and it is much faster with a warm cache.

* I don't think we need watch mode for packages (i.e. `build-packages:watch`). In my experiments, if you run webpack in watch mode and change a `.ts` file in a package, webpack will rebuild as expected. Also, if you change a type in a TS package, VSCode will pick the change, no need to rebuild the types. However, cleaning up watch mode is left for a future PR to avoid scope creep.

### Testing instructions

* Verify tests are green, ICFY doesn't show a crazy diff in bundle size and smoke test calypso.live (in Chrome and IE11), specially areas that have been moved to TS packages (eg: language picker, checkout, shopping cart)

* Checkout this branch and run `yarn` twice. The second time (deps already installed, nothing to build) should finish in a couple of seconds.
